[may18] Update SetManipulationMode to public in TwoHandManipulatable

cre8ivepark opened the pull request:
Overview
---
Update private SetManipulationMode to public in TwoHandManipulatable.cs. This allows the user change manipulation mode option in real-time. (possibly through menu buttons or voice command) Currently, it is only supported in the editor's inspector panel.
![2018-04-20 14_27_20-unity 2017 2 1p2 64bit - uxdemoscene unity - mrtk-demo - universal windows pla](https://user-images.githubusercontent.com/13754172/39074623-055d4fb4-44a7-11e8-9219-c0a0f1d7687d.png)

Separated enum TwoHandManipulatableMode into a new file.

FYI, supported manipulation mode:
- Scale
- Rotate
- Move, Scale
- Rotate, Scale
- Move, Rotate, Scale

Changes
---
- Fixes: #1994

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Utilities/Interactions/TwoHandManipulatableMode.cs:
```diff
@@ -0,0 +1,17 @@
+    /// <summary>
+    /// Sorting type for collections
+    /// </summary>
+    public enum TwoHandManipulatableMode
```
We should probably update this to use the `[Flags]` attribute so we can do proper bit ops on this enum. It'll make working with that switch statement cleaner too.

StephenHodgson commented:
Is this reflecting what's in the Dev branch or will this need to be ported over as well.

I'm concerned we're fixing things in one place and not another, and the branches are going to get out of sync.

cre8ivepark commented:
@StephenHodgson Definitely. I'll make sure to submit same updates to Dev branch too. 
Currently, we have #1849, #1914 and #1936 on the list for the Dev branch.

StephenHodgson commented:
In the future, let's put the fix in dev first, then port to master.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Utilities/Interactions/TwoHandManipulatable.cs:
```diff
@@ -47,6 +47,7 @@ public BoundingBox BoundingBoxPrefab
 
         [SerializeField]
         [Tooltip("What manipulation will two hands perform?")]
+        [EnumFlags]
```
Not sure this is needed.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Utilities/Interactions/TwoHandManipulatableMode.cs:
```diff
@@ -1,17 +1,19 @@
+        Rotate = (1 << 1),
+        MoveScale = (1 << 2),
+        RotateScale = (1 << 3),
+        MoveRotateScale = (1 << 4),
```
I'm suggesting doing something like this:

```
[Flags]
     public enum TwoHandManipulatableMode
{
    None = 1<<0,
    Move = 1<<1,
    Rotate = 1<<2,
    Scale = 1<<3,
}
```

Then in the switch statement, it'll actually get updated to something else that handles it correctly.

StephenHodgson commented on Assets/HoloToolkit/Input/Scripts/Utilities/Interactions/TwoHandManipulatable.cs:
```diff
@@ -255,19 +244,19 @@ private void UpdateStateMachine()
                                 break;
-                            case TwoHandedManipulation.MoveRotateScale:
+                            case TwoHandManipulatableMode.MoveRotateScale:
                                 newState = State.MovingRotatingScaling;
```
Why doesn't state and TwoHandManipulatableMode use the same enum?

StephenHodgson replied:
If they did you could shorten this whole switch statement to just be:
```
newState = TwoHandManipulatableMode;
```

cre8ivepark replied:
@StephenHodgson I would recommend a separate PR for this change since I don't know the details about states handling cases. The scope of this PR is making the manipulation mode as public.

cre8ivepark commented:
@StephenHodgson, one question - have you tried using **[EnumFlags]**? 
Updated enum file with [Flag] attribute and added [EnumFlags] attribute, but the editor does not show the multi-selectable combinations. Does it require additional code to get automatically generated combinations in the editor's Inspector panel? I don't see an example in MRTK.

In TwoHandManipulatableMode.cs:

    [Flags]
    public enum TwoHandManipulatableMode
    {
        Move = (1 << 0),
        Rotate = (1 << 1),
        Scale = (1 << 2),
    }

And in TwoHandManipulatable.cs:

    [EnumFlags]
    public TwoHandManipulatableMode ManipulationMode = TwoHandManipulatableMode.Scale;

![2018-04-30 15_30_48-mrtk-original - microsoft visual studio](https://user-images.githubusercontent.com/13754172/39453634-9fbab7b8-4c8b-11e8-8189-f9d4b975dc5f.png)

StephenHodgson commented:
I believe the checkbox means it's multi selectable.

cre8ivepark commented:
@StephenHodgson Unfortunately it is not :( it is single select dropdown..

StephenHodgson commented:
That's with and without the `[EnumFlags]`?

cre8ivepark commented:
@StephenHodgson Yes, somehow, [EnumgFlags] does not show combinations.
It seems like, still we have to manually add combinations - http://www.alanzucconi.com/2015/07/26/enum-flags-and-bitwise-operators/

StephenHodgson commented:
Pretty sure all you need is `[Flags]` and the default inspector should handle it properly.

Again, not sure `[EnumFlags]` is needed. We do something similar in other inspectors. I could do a quick write-up on your branch if you'd like.

cre8ivepark commented:
@StephenHodgson Thanks! It would be appreciated if you can take a quick look. According to some posts and videos, it seems like [EnumFlags] exposes the UI for automatically generated combinations with multi-selectable buttons. https://www.youtube.com/watch?v=xUFCjIsqaZk

StephenHodgson commented:
Looks like then `[EnumFlags]` attribute in the MRTK is from the MRDL stuff and that class doesn't have a custom editor for it.

I was under the impression that you could use masked enums without custom inspectors, but I was mistaken. Sorry about that.  Ended up writing the editor for the Manipulation class anyway.

StephenHodgson commented:
Change Request for this PR: https://github.com/cre8ivepark/MixedRealityToolkit-Unity/pull/1

cre8ivepark commented:
Thanks! @StephenHodgson 
I have updated example objects' options in the scene to match existing labels.
![2018-05-02 10_48_05-unity 2017 2 1p2 64bit - twohandmanipulationtest unity - stephen - universal w](https://user-images.githubusercontent.com/13754172/39540221-dba4600a-4df6-11e8-9fd6-b264deb1cd5d.png)

StephenHodgson commented:
@cre8ivepark can we also make sure we address #2051 ?

davidkline-ms commented on Assets/HoloToolkit/Input/Scripts/Utilities/Interactions/TwoHandManipulatable.cs:
```diff
@@ -132,9 +106,9 @@ private bool ShowBoundingBox
         /// SetManipulationMode
         /// </summary>
-        private void SetManipulationMode(TwoHandedManipulation mode)
+        public void SetManipulationMode(ManipulationMode mode)
```
Shouldn't this be a property?

Also, there is no GetManipuationMode… client code should be able to see what the current value is as well as set a new one.

StephenHodgson replied:
That's addressed in https://github.com/cre8ivepark/MixedRealityToolkit-Unity/pull/2

StephenHodgson replied:
I made it a property and marked this method obsolete.
